gax/src/main/java/com/google/api/gax/batching/BatcherImpl.java

```diff
@@ -251,26 +252,30 @@ private Batch(
 
     void add(ElementT element, SettableApiFuture<ElementResultT> result) {
       builder.add(element);
-      results.add(result);
+      entries.add(
+          BatchEntry.<ElementT, ElementResultT>newBuilder()
+              .setElement(element)
+              .setResultFuture(result)
+              .build());
       elementCounter++;
       byteCounter += descriptor.countBytes(element);
     }
 
     void onBatchSuccess(ResponseT response) {
       try {
-        descriptor.splitResponse(response, results);
-        batcherStats.recordBatchElementsCompletion(results);
+        descriptor.splitResponse(response, entries);
+        batcherStats.recordBatchElementsCompletion(entries);
       } catch (Exception ex) {
         onBatchFailure(ex);
       }
     }
 
     void onBatchFailure(Throwable throwable) {
       try {
-        descriptor.splitException(throwable, results);
+        descriptor.splitException(throwable, entries);
       } catch (Exception ex) {
-        for (SettableApiFuture<ElementResultT> result : results) {
-          result.setException(ex);
+        for (BatchEntry<ElementT, ElementResultT> batchEntry : entries) {
+          batchEntry.getResultFuture().setException(ex);
```

Comment:
is it possible for `getResultFuture()` to return null?

Reply:
nope. AutoValue will refuse to build the object if a field that isnt annotated with `@nullable` is unset. however you do bring up a good point...why is element nullable. I added a separate comment to discuss that